### DISCUSSION TITLE
How to custom openID connect mappers for client scopes?

### DISCUSSION THREAD

**Discussion by d02540315** [keycloak/keycloak#14708] [Ideas]
Hi, I'm evaluating Keycloak as our on-prem IdP but have some questions. Thank you!

1. How to change the keyword "scope" to "scp" since we have legacy code expecting "scp" from Idp.
2. How to return client scopes (space delimited string) as an Arrary? Our legacy code expects "scp" as an Array from Idp.
3. Is there any built-in mapper I can use to achieve the goals mentioned in 1 and 2 without writing my own mapper?
4. I'm googling a lot these days. It seems legacy version of KC can write custom Java code to implement a custom mapper, but latest version of KC (19) is now talking about use JAR (with js files) to handle the stituation without too much documentation. Can someone provide some thoughts please? 

![image](https://user-images.githubusercontent.com/5162421/193677037-8c7288d7-1fc1-40f8-9be4-a81b2cd63e11.png)

**Comment by sschu** [comment]
I think there is no OOTB way to do that. However, using a custom mapper you can still implement this using Java code only.

**Comment by jbman** [comment]
With some extra setup effort it is possible to get the a scp claim based on a client and a client role for each scope value:

* Create a client scp-profile to define role "profile" and client scp-email to define role "email"
* Add these roles to all users
* Add mappers to client scopes "email" and client scopes "profile" like shown below

Result: The client roles are mapped into claim "scp" into an array.

![image](https://user-images.githubusercontent.com/789496/195058442-e54ea271-d5d4-4b46-967d-9fdcf89fd1d9.png)
